mempool/reactor.go

```diff
@@ -218,26 +215,24 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
 		time.Sleep(PeerCatchupSleepIntervalMS * time.Millisecond)
 	}
 
+	iter := memR.mempool.NewIterator()
+	var entry Entry
 	for {
 		// In case of both next.NextWaitChan() and peer.Quit() are variable at the same time
 		if !memR.IsRunning() || !peer.IsRunning() {
 			return
 		}
 
-		// This happens because the CElement we were looking at got garbage
-		// collected (removed). That is, .NextWait() returned nil. Go ahead and
-		// start from the beginning.
-		if next == nil {
-			select {
-			case <-memR.mempool.TxsWaitChan(): // Wait until a tx is available
-				if next = memR.mempool.TxsFront(); next == nil {
-					continue
-				}
-			case <-peer.Quit():
-				return
-			case <-memR.Quit():
-				return
+		select {
+		case entry = <-iter.WaitNextCh():
+			// If the entry we were looking at got garbage collected (removed), try again.
+			if entry == nil {
```

Comment:
If we reach the end, do we also retry here?

Reply:
In principle, yes, that's what would happen.

Here, `entry` is nil either because `Clist.Front` or `CElement.Next` returned nil. If `Next` returns nil because it's at the end of the list and the list is not empty, we shouldn't set the cursor to nil because we will start iterating from the beginning of the list, sending the same transactions more than once. The other reason I see that both `Front` and `Next` return nil is when the entry got removed just before calling these functions. I've seen these scenarios happen in unit tests, but they are difficult to reproduce.

I think it's better to investigate these scenarios in another issue; this PR should be just a refactoring.